libraries/adaptive-expressions/src/functionUtils.ts

```diff
@@ -0,0 +1,1033 @@
+            const instanceMap: Map<string, any> = instance as Map<string, any>;
+            value = instanceMap.get(property);
+            if (value === undefined) {
+                const prop: string = Array.from(instanceMap.keys()).find((k: string): boolean => k.toLowerCase() === property.toLowerCase());
+                if (prop !== undefined) {
+                    value = instanceMap.get(prop);
+                }
+            }
+        } else {
+            const prop: string = Object.keys(instance).find((k: string): boolean => k.toLowerCase() === property.toLowerCase());
+            if (prop !== undefined) {
+                value = instance[prop];
+            }
+        }
+
+        return {value, error};
+    }
+
+    /**
+     * Lookup a property in IDictionary, JObject or through reflection.
+     * @param instance Instance with property.
+     * @param property Property to lookup.
+     * @returns Value and error information if any.
+     */
+    public static accessIndex(instance: any, index: number): {value: any; error: string} {
+        // NOTE: This returns null rather than an error if property is not present
+        if (instance === null || instance === undefined) {
+            return {value: undefined, error: undefined};
+        }
+
+        let value: any;
+        let error: string;
+
+        if (Array.isArray(instance)) {
+            if (index >= 0 && index < instance.length) {
+                value = instance[index];
+            } else {
+                error = `${index} is out of range for ${instance}`;
+            }
+        } else {
+            error = `${instance} is not a collection.`;
+        }
+
+        return {value, error};
+    }
+
+    /**
+     * transform a timestamp into another with customized function.
+     * @param timeStamp Original time stamp.
+     * @param transform Transform function.
+     * @returns new timestamp and error.
+     */
+    public static parseTimestamp(timeStamp: string, transform?: (arg0: Date) => any): {value: any; error: string} {
+        let value: any;
+        const error: string = this.verifyISOTimestamp(timeStamp);
+        if (!error) {
+            value = transform !== undefined ? transform(new Date(timeStamp)) : timeStamp;
+        }
+
+        return {value, error};
+    }
+
+    /**
+     * timestampFormatter, to convert C# datetime to moment js format.
+     * @param formatter C# datetime format
+     */
+    public static timestampFormatter(formatter: string): string {
+        let result = formatter;
+        try {
+            result = convertCSharpDateTimeToMomentJS(formatter);
+        } catch (e) {
+            // do nothing
+        }
+
+        return result;
+    }
+
+    /**
+     * Transform C# duration and unit into js duration and unit
+     * @param duration C# duration
+     * @param cSharpStr C# unit.
+     */
+    public static timeUnitTransformer(duration: number, cSharpStr: string): {duration: number; tsStr: string} {
```

Comment:
>{duration: number; tsStr: string} [](start = 76, length = 33)

If this and other structures are used in more than one place turn them into TS types.

Reply:
There just three places using this data type. I think there is no need to add an additional type to describe this data format.
Otherwise, typescript recommends using `interface` instead of `type` to describe a data format.